Content.Shared/Chemistry/Components/Solution.cs

```diff
@@ -82,38 +130,152 @@ public bool CanAddSolution(Solution solution)
-        ///     The total heat capacity of all reagents in the solution.
-        /// </summary>
-        [ViewVariables] private float _heatCapacity;
+        ///     Updates the heat capacity after adding or taking from the solution.
+        /// </summary
+        public void UpdateHeatCapacity(IPrototypeManager? protoMan = null)
+        {
+            IoCManager.Resolve(ref protoMan);
+            HeatCapacity = 0;
+            foreach(var(reagent, quantity) in Contents)
+            {
+                var specificHeat = protoMan.Index<ReagentPrototype>(reagent.Prototype).SpecificHeat;
+                HeatCapacity += (float)quantity * specificHeat;
+            }
+        }
 
         /// <summary>
-        ///     If true, then <see cref="_heatCapacity"/> needs to be recomputed.
-        /// </summary>
-        [ViewVariables] private bool _heatCapacityDirty = true;
+        ///     Returns the total thermal energy the solution holds.
+        /// </summary
+        public float GetThermalEnergy()
+        {
+            return HeatCapacity * Temperature;
+        }
 
-        public void UpdateHeatCapacity(IPrototypeManager? protoMan)
+        /// <summary>
+        ///     Sets the temperature of the solution.
+        /// </summary
+        public void SetAbsoluteTemperature(float temperature)
         {
-            IoCManager.Resolve(ref protoMan);
-            DebugTools.Assert(_heatCapacityDirty);
-            _heatCapacityDirty = false;
-            _heatCapacity = 0;
-            foreach (var (reagent, quantity) in Contents)
+            Temperature = temperature;
+        }
+
+        /// <summary>
+        ///     Sets the temperature of the solution according to the amount of heat it should have.
+        /// </summary
+        public void SetTemperature(float thermalEnergy, IPrototypeManager? protoMan = null)
+        {
+            if(HeatCapacity == 0)
+                return;
+            Temperature = thermalEnergy / HeatCapacity; 
+            ValidateTemperature(protoMan);
+        }
+
+        /// <summary>
+        ///     Adjusts the temperature of the solution according the the amount of heat its given or taken.
+        /// </summary
+        public void AdjustTemperature(float thermalEnergy, IPrototypeManager? protoMan = null)
+        {
+            if(HeatCapacity == 0)
+                return;
+
+            if(thermalEnergy > 0)
             {
-                _heatCapacity += (float) quantity *
-                                 protoMan.Index<ReagentPrototype>(reagent.Prototype).SpecificHeat;
+                MeltingLatentHeat = 0;
+                ShouldFreezeInstantly = false;
             }
+            else
+            {
+                BoilingLatentHeat = 0;
+                ShouldBoilInstantly = false;
+            }
+
+            Temperature += thermalEnergy / HeatCapacity;
+            ValidateTemperature(protoMan);
         }
 
-        public float GetHeatCapacity(IPrototypeManager? protoMan)
```

Comment:
Could you please explain why this API breakage is needed?

Reply:
I think the code looks better if heat capacity is just a variable with private set for HeatCapacity. Also I didn't think _heatCapacityDirty was needed, so no need to check for it when you want to get the heat capacity.

Reply:
the idea is to intentionally defer updating the heat capacity until its required to make modifying solutions faster. This is also why the cached heat capacity wasn't publicly available - so that people didn't use it without first checking that the heat capacity is up to date

Reply:
using protoman to get a prototype, even if we don't resolve it, is really expensive then?

Reply:
it is comparatively expensive for something that needs to happen very frequently. Solutions are used by a lot of other component/systems and there are many things that modify them every tick.